### PR TITLE
OData Migration

### PR DESCRIPTION
**Pull request by groovyfeng**

### CODE REVIEW

File: sln/ODataMigration.sln [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 15
+VisualStudioVersion = 15.0.28307.645
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.Extensions.OData.Migration", "..\src\Microsoft.Extensions.OData.Migration\Microsoft.Extensions.OData.Migration.csproj", "{4575BD2C-1117-4CA2-8500-24BBB649FAFB}"
```
Comment (xuzhg) [inline sln/ODataMigration.sln]: >Microsoft.Extensions.OData.Migration [](start = 53, length = 36)

Microsoft.Extensions.* is resolved by other teams.

So, we have to change it. Please change it as;

Microsoft.OData.Extensions.Migration

File: src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+    using Microsoft.AspNet.OData.Routing;
+    using Microsoft.OData.Edm;
+    using Microsoft.Spatial;
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.IO;
+    using System.Linq;
+    using System.Reflection;
+    using System.Xml.Linq;
```
Comment (xuzhg) [inline src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs]: Please sort the using. 

1. Move the System.* at head
2. Alphabet the others

File: src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+            {
+                return String.Format(
+                    CultureInfo.InvariantCulture,
+                    "{0}Of{1}",
+                    type.Name.Replace('`', '_'),
+                    String.Join("_", type.GetGenericArguments().Select(t => MangleClrTypeName(t))));
+            }
+        }
+
+
```
Comment (xuzhg) [inline src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs]: remove unused the lines

File: src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+        /// <summary>
+        /// Finds a V4 type definition by the name of a V3 type definition
+        /// </summary>
+        /// <param name="model">V4 model to search</param>
+        /// <param name="t">V3 type</param>
+        /// <returns>Equivalent V4 Type or null if not found</returns>
+        public static IEdmType GetV4Definition(this IEdmModel model, Data.Edm.IEdmType t)
```
Comment (xuzhg) [inline src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs]: >GetV4Definition [](start = 31, length = 15)

should we use "ODataVersion" to specify the version?

File: src/Microsoft.Extensions.OData.Migration/Microsoft.Extensions.OData.Migration.csproj [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
+    <Version>0.9.2-preview</Version>
```
Comment (xuzhg) [inline src/Microsoft.Extensions.OData.Migration/Microsoft.Extensions.OData.Migration.csproj]: >0.9.2-preview [](start = 13, length = 13)

the version is by design?

**Comment by xuzhg** [comment]
Don't include the .dll into the commitment, See System.Web.Http.OData.dll or others in the test project

File: test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/InlineCountTranslationTest.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        private static readonly Uri serviceRoot = new Uri("http://localhost:80/");
+        private static readonly ODataMigrationMiddleware middleware = TestModelProvider.ODataSvcSampleMiddleware(serviceRoot);
+
+        [Theory]
+        [MemberData(nameof(InlineCountTranslationQueries))]
+        public void TestInlineCountTranslations(string name, string testQuery, string expectedQuery)
+        {
+            Uri result = middleware.TranslateUri(new Uri(serviceRoot, testQuery));
+            Uri expected = new Uri(serviceRoot, expectedQuery == "IS_SAME" ? testQuery : expectedQuery);
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/InlineCountTranslationTest.cs]: >expectedQuery == "IS_SAME" [](start = 48, length = 26)

It seems there's not such test case

File: test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/InlineCountTranslationTest.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    using Xunit;
+
+    public class InlineCountTranslationTest
+    {
+        private static readonly Uri serviceRoot = new Uri("http://localhost:80/");
+        private static readonly ODataMigrationMiddleware middleware = TestModelProvider.ODataSvcSampleMiddleware(serviceRoot);
+
+        [Theory]
+        [MemberData(nameof(InlineCountTranslationQueries))]
+        public void TestInlineCountTranslations(string name, string testQuery, string expectedQuery)
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/InlineCountTranslationTest.cs]: >name [](start = 55, length = 4)

where's using "name"? same for other test cases

If that's only for the comments, just add them as comments into the test case. for example: 

{  new object[] { "Products?$inlinecount=allpages", "Products?$count=true" } }, // All pages should become true

File: test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/BodyTranslationTest.cs [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+            Assert.Equal(HttpStatusCode.OK, customersResponse.StatusCode);
+        }
+        
+        [Fact]
+        public async void GetSimpleResourceSetAsV3HasV3Types()
+        {
+            HttpResponseMessage response = await Get(OrderDetailsBaseUrl, AddODataV3Header);
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            string content = await response.Content.ReadAsStringAsync();
+            Assert.Contains(@"""1000""", content); // Check for 10,000 because it is a long value in the generated OrderDetails
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/BodyTranslationTest.cs]: >10,000 [](start = 64, length = 6)

make the comment correct

File: src/Microsoft.Extensions.OData.Migration/README.md [hunk 9]
```diff
@@ -0,0 +1,262 @@
+﻿# Introduction
+
+The OData Migration library provides ASP.NET Core 2.2+ OData V4 services with the capability of handling and responding to OData V3 requests.  An OData V4 service that uses the OData Migration Library may appear to an OData V3 client as the equivalent V3 service.  This may be useful to you if you have migrated your service from OData V3 to OData V4, but wish to bridge the gap between your newly migrated V4 service and your OData V3 clients.
+
+# Usage
```
Comment (xuzhg) [inline src/Microsoft.Extensions.OData.Migration/README.md]: please prepare the similar md file and file an PR to 
[https://github.com/MicrosoftDocs/OData-docs/pulls](https://github.com/MicrosoftDocs/OData-docs/pulls)

File: src/Microsoft.Extensions.OData.Migration/MigrationExtension.cs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    public static class MigrationExtension
+    {
+        /// <summary>
+        /// Extension method to use V3 to V4 translation middleware.
+        /// </summary>
+        /// <param name="builder">IApplicationBuilder that will use translation middleware</param>
+        /// <param name="v3Edmx">V3 edmx to send back when requested for metadata</param>
+        /// <param name="v4Model">Required V4 model to validate translated URI</param>
+        /// <returns>builder now using migration middleware</returns>
+        public static IApplicationBuilder UseODataMigration(this IApplicationBuilder builder,
```
Comment (xuzhg) [inline src/Microsoft.Extensions.OData.Migration/MigrationExtension.cs]: > IApplicationBuilder UseODataMigration(this IApplicationBuilder builder, [](start = 21, length = 72)

how can customer use the whole extensions methods to build his service? Would you please create a ASP.NET CORE application sample to play with them?

File: src/Microsoft.Extensions.OData.Migration/Microsoft.Extensions.OData.Migration.csproj [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    <Authors>Microsoft</Authors>
```
Comment (xuzhg) [inline src/Microsoft.Extensions.OData.Migration/Microsoft.Extensions.OData.Migration.csproj]: These setting maybe not work with "<import project="..\Build.props" />, Just FYI.

File: sln/ODataMigration.sln (hunk 1 above)
Comment (groovyfeng) [reply sln/ODataMigration.sln]: changed

File: src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs (hunk 3 above)
Comment (groovyfeng) [reply src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs]: removed

File: src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs (hunk 2 above)
Comment (groovyfeng) [reply src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs]: changed

File: src/Microsoft.Extensions.OData.Migration/Microsoft.Extensions.OData.Migration.csproj (hunk 5 above)
Comment (groovyfeng) [reply src/Microsoft.Extensions.OData.Migration/Microsoft.Extensions.OData.Migration.csproj]: No, it's used for internal consumption in Intune. I will change it to 1.0.0 unless you have other recommendations.

File: test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/InlineCountTranslationTest.cs (hunk 6 above)
Comment (groovyfeng) [reply test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/InlineCountTranslationTest.cs]: Right, not needed. Removed the check.

File: test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/InlineCountTranslationTest.cs (hunk 7 above)
Comment (groovyfeng) [reply test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/InlineCountTranslationTest.cs]: change name to comments

File: test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/BodyTranslationTest.cs (hunk 8 above)
Comment (groovyfeng) [reply test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/BodyTranslationTest.cs]: fixed

File: src/Microsoft.Extensions.OData.Migration/README.md (hunk 9 above)
Comment (groovyfeng) [reply src/Microsoft.Extensions.OData.Migration/README.md]: Will do

File: src/Microsoft.Extensions.OData.Migration/MigrationExtension.cs (hunk 10 above)
Comment (groovyfeng) [reply src/Microsoft.Extensions.OData.Migration/MigrationExtension.cs]: The extension is being used in Intune's StatelessConnectorService: https://msazure.visualstudio.com/One/_git/Intune-Svc-RACerts?path=%2Fsrc%2FServices%2FConnectorService%2FSource%2FStartup.cs&version=GBdevelop&line=144&lineStyle=plain&lineEnd=144&lineStartColumn=13&lineEndColumn=79

I will make sure it's doc'd clearly in the md file.

File: src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs (hunk 4 above)
Comment (groovyfeng) [reply src/Microsoft.Extensions.OData.Migration/EdmExtensions.cs]: We always want to get the OData4 definition, so there is no need to specify the version.

**Pull request merged**